### ISSUE TITLE
swprintf %s format for wide strings

### ISSUE THREAD

**Issue by OZ1** [OpenVPN/openvpn-gui#397]
I'm trying to build the project using MSYS2 following BUILD.rst setup instruction.
In the resulting binary I get some strings truncated after the first character. The cause is that for unicode strings "%s" formatting is used, which is perfectly fine in MSVC. However MinGW sprintf functions expect "%ls" for wide strings. How did it suppose to work? How do you compile the app to include MSVC-style swprintfs? What steps are missing from BUILD.rst?

**Comment by OZ1** [comment]
Maybe _swnprintf should be used instead?

**Comment by selvanair** [comment]
Our official builds are made using mingw-w64 cross compiler on Linux. The code should work as is without exhibiting any string truncations. What version of mingw-w64?

As mingw targets windows runtime, "%s" in swprintf gets interpreted as corresponding to wide string, and "%S" as narrow string. That's Windows convention that predates C99 which requires %ls and %hs. Read about it here. https://devblogs.microsoft.com/oldnewthing/20190830-00/?p=102823

That said, I don't mind changing %s to %ls and %S (if any) to %hs to be explicit as (I think) that also works on all modern versions of Windows. 

I don't build on Windows, but the instructions in BUILD.rst look right to me. Here is how to cross-compile on Linux using mingw-w64  --- showing the simpler case of building without OpenSSL by disabling the pkey password change feature.

```
autoreconf -iv
make clean 
./configure --host=x86_64-w64-mingw32 --build=x86_64-pc-linux-gnu --disable-password-change
 make
```

**Comment by OZ1** [comment]
it seems that just changing the functions does the job

**Comment by selvanair** [comment]
> it seems that just changing the functions does the job

As I said swprintf should work as is. Please report your compiler version and documentation that says swprintf behaves differently on that version.

**Comment by OZ1** [comment]
I am building the app in MinGW on Windows.

http://www.cplusplus.com/reference/cwchar/swprintf:
Notice that all format specifiers have the same meaning as in printf; therefore, %lc shall be used to write a wide character (and not %c), as well as %ls shall be used for wide strings (and not %s).
Format specifiers are described here: http://www.cplusplus.com/reference/cstdio/printf/

To reuse a build system, I've inserted some lines in main.c:
`int WINAPI _tWinMain (HINSTANCE hThisInstance,
                    UNUSED HINSTANCE hPrevInstance,
                    UNUSED LPTSTR lpszArgument,
                    UNUSED int nCmdShow)
{
#define THE_STRING L"A string"
wchar_t Buffer[100];
_snwprintf(Buffer, _countof(Buffer), L"The string from _snwprintf is '%s' (should read like '" THE_STRING L"')\t\n", THE_STRING);
MessageBox(0, Buffer, 0, 0);
swprintf(Buffer, _countof(Buffer), L"The string from swprintf is '%s' (should read like '" THE_STRING L"')\r\n", THE_STRING);
MessageBox(0, Buffer, 0, 0);
  MSG messages;            /* Here messages to the application are saved */
  WNDCLASSEX wincl;        /* Data structure for the windowclass */
  DWORD shell32_version;
  BOOL first_instance = TRUE;
`

MessageBox out is the following:
---------------------------
Ошибка
---------------------------
The string from _snwprintf is 'A string' (should read like 'A string')	

---------------------------
ОК   
---------------------------

---------------------------
Ошибка
---------------------------
The string from swprintf is 'A' (should read like 'A string')


---------------------------
ОК   
---------------------------

**Comment by OZ1** [comment]
adding " --host=x86_64-w64-mingw32 --build=x86_64-pc-linux-gnu --disable-password-change" to ./configure does not change anything

**Comment by OZ1** [comment]
> Our official builds are made using mingw-w64 cross compiler on Linux. The code should work as is without exhibiting any string truncations. What version of mingw-w64?
$ uname -a
MINGW64_NT-10.0-18363 OZonov-QEMU 3.1.7-340.x86_64 2020-11-08 12:32 UTC x86_64 Msys
How can I lookup the version of mingw-w64? I see 20190524 for MSYS2 64 bit in Add and Remove Programs control panel applet.

**Comment by selvanair** [comment]
> I am building the app in MinGW on Windows.
> http://www.cplusplus.com/reference/cwchar/swprintf:

You are referring to irrelevant documentation. This is Windows which doesn't always follow the C standard. Refer to MSDN. Note that mingw targets Windows runtime, not cygwin.

The result of the example you posted seems to indicate your compiler is not targeting Windows runtime for stdio or not building with mingw. Otherwise both those message boxes should display the same text.

That said, as noted before, I'm open to changing %s to %ls to be explicit. No need to change to _snwprintf.

**Comment by OZ1** [comment]
> That said, as noted before, I'm open to changing %s to %ls to be explicit. No need to change to _snwprintf.

There are 6 calls to swprintf, 8 calls to _snwprintf throught the code base and 37 calls to _vsntprintf via _snprintf_0. It seems logical to me to change the minority of calls to follow the remainig style.
"%ls" looks awkward to me as Windows developer. I don't feel '%ls' a good comitee descision. Microsoft did better in this field.

> You are referring to irrelevant documentation. This is Windows which doesn't always follow the C standard. Note that mingw targets Windows runtime, not cygwin. Refer to MSDN here and here.

This is not an ordidary msbuild build though. I see gcc being invoked, the beast unknown for me. This why I didn't refer to MSDN.
```
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal
autoreconf: configure.ac: tracing
autoreconf: configure.ac: not using Libtool
autoreconf: running: /usr/bin/autoconf
autoreconf: running: /usr/bin/autoheader
autoreconf: running: automake --no-force
autoreconf: Leaving directory `.'
test -z "openvpn-gui.exe" || rm -f openvpn-gui.exe
rm -f *.o
configure: loading site script /mingw64/etc/config.site
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking build system type... x86_64-w64-mingw32
checking host system type... x86_64-w64-mingw32
checking whether make supports the include directive... yes (GNU style)
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.exe
checking for suffix of executables... .exe
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... gcc3
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking minix/config.h usability... no
checking minix/config.h presence... no
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking for gcc option to accept ISO C99... none needed
checking for windres... windres
checking how to run the C preprocessor... gcc -E
checking whether ln -s works... no, using cp -pR
checking whether make sets $(MAKE)... (cached) yes
checking for pkg-config... /mingw64/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for OPENSSL_CRYPTO... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating config.h
config.status: executing depfiles commands
make  all-am
make[1]: Entering directory '/c/openvpn-gui'
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT main.o -MD -MP -MF .deps/main.Tpo -c -o main.o main.c
main.c: In function 'GetDllVersion':
main.c:840:26: warning: cast between incompatible function types from 'FARPROC' {aka 'long long int (*)()'} to 'HRESULT (*)(DLLVERSIONINFO *)' {aka 'long int (*)(DLLVERSIONINFO *)'} [-Wcast-function-type]
  840 |         pDllGetVersion = (DLLGETVERSIONPROC)GetProcAddress(hinstDll,
      |                          ^
mv -f .deps/main.Tpo .deps/main.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT openvpn.o -MD -MP -MF .deps/openvpn.Tpo -c -o openvpn.o openvpn.c
openvpn.c: In function 'UserAuthDialogFunc':
openvpn.c:532:16: warning: this statement may fall through [-Wimplicit-fallthrough=]
  532 |             if (HIWORD(wParam) == EN_UPDATE)
      |                ^
openvpn.c:538:9: note: here
  538 |         case ID_EDT_AUTH_PASS:
      |         ^~~~
mv -f .deps/openvpn.Tpo .deps/openvpn.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT localization.o -MD -MP -MF .deps/localization.Tpo -c -o localization.o localization.c
mv -f .deps/localization.Tpo .deps/localization.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT tray.o -MD -MP -MF .deps/tray.Tpo -c -o tray.o tray.c
mv -f .deps/tray.Tpo .deps/tray.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT viewlog.o -MD -MP -MF .deps/viewlog.Tpo -c -o viewlog.o viewlog.c
mv -f .deps/viewlog.Tpo .deps/viewlog.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT service.o -MD -MP -MF .deps/service.Tpo -c -o service.o service.c
mv -f .deps/service.Tpo .deps/service.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT options.o -MD -MP -MF .deps/options.Tpo -c -o options.o options.c
mv -f .deps/options.Tpo .deps/options.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT passphrase.o -MD -MP -MF .deps/passphrase.Tpo -c -o passphrase.o passphrase.c
mv -f .deps/passphrase.Tpo .deps/passphrase.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT proxy.o -MD -MP -MF .deps/proxy.Tpo -c -o proxy.o proxy.c
mv -f .deps/proxy.Tpo .deps/proxy.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT registry.o -MD -MP -MF .deps/registry.Tpo -c -o registry.o registry.c
mv -f .deps/registry.Tpo .deps/registry.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT scripts.o -MD -MP -MF .deps/scripts.Tpo -c -o scripts.o scripts.c
mv -f .deps/scripts.Tpo .deps/scripts.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT manage.o -MD -MP -MF .deps/manage.Tpo -c -o manage.o manage.c
mv -f .deps/manage.Tpo .deps/manage.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT misc.o -MD -MP -MF .deps/misc.Tpo -c -o misc.o misc.c
misc.c: In function 'Base32Decode':
misc.c:136:26: warning: operand of '?:' changes signedness from 'long long int' to 'size_t' {aka 'long long unsigned int'} due to unsignedness of other operand [-Wsign-compare]
  136 |     DWORD in_len = end ? (end - input) : strlen(input);
      |                          ^~~~~~~~~~~~~
mv -f .deps/misc.Tpo .deps/misc.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT openvpn_config.o -MD -MP -MF .deps/openvpn_config.Tpo -c -o openvpn_config.o openvpn_config.c
mv -f .deps/openvpn_config.Tpo .deps/openvpn_config.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT access.o -MD -MP -MF .deps/access.Tpo -c -o access.o access.c
mv -f .deps/access.Tpo .deps/access.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT save_pass.o -MD -MP -MF .deps/save_pass.Tpo -c -o save_pass.o save_pass.c
save_pass.c: In function 'save_encrypted_data':
save_pass.c:107:25: warning: passing argument 1 of 'crypt_protect' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
  107 |     len = crypt_protect(password, len, entropy, &out);
      |                         ^~~~~~~~
save_pass.c:21:21: note: expected 'BYTE *' {aka 'unsigned char *'} but argument is of type 'const void *'
   21 | crypt_protect(BYTE *data, int szdata, char *entropy, BYTE **out)
      |               ~~~~~~^~~~
save_pass.c: In function 'recall_encrypted':
save_pass.c:187:52: warning: passing argument 3 of 'recall_encrypted_data' from incompatible pointer type [-Wincompatible-pointer-types]
  187 |     len = recall_encrypted_data(config_name, name, &out);
      |                                                    ^~~~
      |                                                    |
      |                                                    BYTE ** {aka unsigned char **}
save_pass.c:163:75: note: expected 'void **' but argument is of type 'BYTE **' {aka 'unsigned char **'}
  163 | recall_encrypted_data(const WCHAR *config_name, const WCHAR *name, void **out)
      |                                                                    ~~~~~~~^~~
save_pass.c: In function 'recall_encrypted_1':
save_pass.c:214:52: warning: passing argument 3 of 'recall_encrypted_data' from incompatible pointer type [-Wincompatible-pointer-types]
  214 |     len = recall_encrypted_data(config_name, name, &out);
      |                                                    ^~~~
      |                                                    |
      |                                                    BYTE ** {aka unsigned char **}
save_pass.c:163:75: note: expected 'void **' but argument is of type 'BYTE **' {aka 'unsigned char **'}
  163 | recall_encrypted_data(const WCHAR *config_name, const WCHAR *name, void **out)
      |                                                                    ~~~~~~~^~~
mv -f .deps/save_pass.Tpo .deps/save_pass.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT env_set.o -MD -MP -MF .deps/env_set.Tpo -c -o env_set.o env_set.c
mv -f .deps/env_set.Tpo .deps/env_set.Po
gcc -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -municode -g -O2 -pedantic -Wall -Wextra -MT echo.o -MD -MP -MF .deps/echo.Tpo -c -o echo.o echo.c
mv -f .deps/echo.Tpo .deps/echo.Po
windres -DHAVE_CONFIG_H -I.  -IC:/msys64/mingw64/include  -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_WIN32_WINNT=_WIN32_WINNT_VISTA -DWINVER=_WIN32_WINNT -i res/openvpn-gui-res.rc -o openvpn-gui-res.o
gcc -municode -g -O2 -pedantic -Wall -Wextra -mwindows -Wl,--nxcompat,--dynamicbase -Wl,--high-entropy-va -o openvpn-gui.exe main.o openvpn.o localization.o tray.o viewlog.o service.o options.o passphrase.o proxy.o registry.o scripts.o manage.o misc.o openvpn_config.o access.o save_pass.o env_set.o echo.o openvpn-gui-res.o -LC:/msys64/mingw64/lib -lcrypto  -lws2_32 -lcomctl32 -lwinhttp -lwtsapi32 -lcrypt32 -lnetapi32 -lole32 -lshlwapi -lsecur32
make[1]: Leaving directory '/c/openvpn-gui'
```

**Comment by selvanair** [comment]
As we are targeting Windows runtime for stdio, MSDN is the right reference and as I said before our use of swprintf is correct. Errors you getcould signal larger problems like using an unsupported compiler. Someone who uses MSYS2 may be able to help better.

What does gcc -v show? Also check whether you have multiple gcc executables installed and whether the PATH is correct so that the one in mingw64/bin is getting picked.  You have to use the one installed by `pacman -S mingw-w64-x86_64-gcc`, not `pacman -S gcc`.

**Comment by OZ1** [comment]
> What does gcc -v show?

$ gcc -v
Using built-in specs.
COLLECT_GCC=C:\msys64\mingw64\bin\gcc.exe
COLLECT_LTO_WRAPPER=C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ../gcc-10.2.0/configure --prefix=/mingw64 --with-local-prefix=/mingw64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/mingw64/x86_64-w64-mingw32/include --libexecdir=/mingw64/lib --enable-bootstrap --with-arch=x86-64 --with-tune=generic --enable-languages=c,lto,c++,fortran,ada,objc,obj-c++,jit --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-filesystem-ts=yes --enable-libstdcxx-time=yes --disable-libstdcxx-pch --disable-libstdcxx-debug --disable-isl-version-check --enable-lto --enable-libgomp --disable-multilib --enable-checking=release --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --disable-plugin --with-libiconv --with-system-zlib --with-gmp=/mingw64 --with-mpfr=/mingw64 --with-mpc=/mingw64 --with-isl=/mingw64 --with-pkgversion='Rev6, Built by MSYS2 project' --with-bugurl=https://github.com/msys2/MINGW-packages/issues --with-gnu-as --with-gnu-ld --with-boot-ldflags='-pipe -Wl,--dynamicbase,--high-entropy-va,--nxcompat,--default-image-base-high -Wl,--disable-dynamicbase -static-libstdc++ -static-libgcc' 'LDFLAGS_FOR_TARGET=-pipe -Wl,--dynamicbase,--high-entropy-va,--nxcompat,--default-image-base-high' --enable-linker-plugin-flags='LDFLAGS=-static-libstdc++\ -static-libgcc\ -pipe\ -Wl,--dynamicbase,--high-entropy-va,--nxcompat,--default-image-base-high\ -Wl,--stack,12582912'
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 10.2.0 (Rev6, Built by MSYS2 project)

**Comment by OZ1** [comment]
> You have to use the one installed by `pacman -S mingw-w64-x86_64-gcc`, not `pacman -S gcc`

Never did none of these as far as I remember. I did exactly as BUILD.rst says.

**Comment by selvanair** [comment]
I checked this using a latest MSYS2  install and could reproduce the behaviour. After some digging it turns out __USE_MINGW_ANSI_STDIO is getting defined as 1 due to some changes in recent versions of mingw. The use of this macro is to replace non-conformant windows stdio routines by custom ones which we have so far avoided. 

Until we understand full implications of this we have to build with that macro explicitly defined as 0.  
Try building like this:
`CFLAGS=-D__USE_MINGW_ANSI_STDIO=0 ./configure` followed by `make`.

PS. 
Something our configure.ac defines _GNU_SOURCE and POSIX_SOURCE which are some of the conditions under which newer mingw forces __USE_MINGW_ANSI_STDIO = 1 unless explicitly defined as 0.

**Comment by selvanair** [comment]
Resolved by #459
